### ISSUE TITLE
How do we tell if the training of SRGAN is stable?

### ISSUE THREAD

**Issue by Jing25** [xinntao/BasicSR#117]
Hi @xinntao,

How can I tell if the training is doing the right thing? Here are some outputs. Does it mean that my training is not stable and I need to adjust my hyperparameters? I use SRGAN.

![image](https://user-images.githubusercontent.com/27107288/57101360-ea861380-6cd5-11e9-9f3d-2722f77f71ad.png)
![image](https://user-images.githubusercontent.com/27107288/57101652-8879de00-6cd6-11e9-991c-09e16cc9770e.png)
![image](https://user-images.githubusercontent.com/27107288/57101885-12c24200-6cd7-11e9-9bd3-797f71db8a20.png)
![image](https://user-images.githubusercontent.com/27107288/57101691-9e879e80-6cd6-11e9-8a18-a1c0102cd24b.png)
![image](https://user-images.githubusercontent.com/27107288/57101721-a8110680-6cd6-11e9-9c2f-a69050428d13.png)

In general, how do I know if the training is relatively stable in super-resolution using GANs? What statistic metrics do I look?

Thank you!

**Comment by xinntao** [comment]
1. Look at the losses. If GAN loss has an abrupt change, usually it indicates there are some problems.
2. I will check the validation images **visually** during training to get to know whether it is stable.

**Comment by Jing25** [comment]
I see. So the strange changes in PSNR doesn't indicate that the training is unstable right?

**Comment by xinntao** [comment]
Yes, PSNR may fluctuate a lot.

**Comment by Jing25** [comment]
Thank you!

**Comment by IceClear** [comment]
Hi, @xinntao, thanks for your awesome project.  During training ESRGAN, I got the same loss curve as hers presented above.  I am still new to GAN. In my view, the D loss and G loss should both decrease and the abrupt change seems very strange. So what may the problem be? Would you give any advice? Should I adjust the learning rate or loss weight or switch to WGAN-GP or anything else? Thanks a lot!

**Comment by xinntao** [comment]
@IceClear 
Do your losses look strange? How about the visual quality of images?
Usually, the problem arises from the loss weight.

**Comment by IceClear** [comment]
@xinntao Thanks for your reply! Actually, I train on my own dataset. I first pretrain G with l1 loss as you suggested. I found the loss decrease very fast, usually after 5k, the loss do not decrease obviously anymore. Then I turn to GAN training. I keep the weight the same as ESRGAN, since I do not know how to set it. I tried to pretrain D for 1k iterations, but it does not work. Now, my loss is as the figure shows below. I only train on the Y channel and as you can see, the val image is full of noise. I t seems the G generate noise and add it to the origin frame. Now, I mainly have two concerns:
(1) Will some GAN improvement work? Such as label smoothing or any other tricks;  Or if  I switch to WGAN-GP instead of RAGAN ?
(2) From the loss, I found at early iterations, the G loss has a slight decrease but become to increase after some iters and the D loss seems to shock and can be 0 sometimes. Does it mean the G is hard to fool the D and the training fails? Maybe the problem is my dataset? I am training on a video enhancement dataset in paper Multi-Frame Quality Enhancement for Compressed Video. So the GT is too hard for the G to learn?
Thanks a lot for your patient reply.
![image](https://user-images.githubusercontent.com/22350795/65007654-69a07f00-d939-11e9-81a1-d8dcc4608e24.png)
![image](https://user-images.githubusercontent.com/22350795/65007669-73c27d80-d939-11e9-8125-9328f0addbbd.png)
![image](https://user-images.githubusercontent.com/22350795/65008099-c7819680-d93a-11e9-8bb3-8268875ea4d6.png)
(This is the test result after 60k iterations, I found the continue training only result in different noise format)

**Comment by IceClear** [comment]
![image](https://user-images.githubusercontent.com/22350795/65008178-031c6080-d93b-11e9-9cd3-76416401e2fa.png)

**Comment by 924175302** [comment]
@IceClear
Hello, I am a newbie. I would like to ask what is your picture based on. Is it a training log? So where is the code? Hope you can answer

**Comment by IceClear** [comment]
They are just the results showed via TensorboardX.

**Comment by 924175302** [comment]
@IceClear
Thank you for your answer. Is the code for presenting the experimental results with TensorboardX in basicSR or did you write it yourself?

**Comment by IceClear** [comment]
> @IceClear
> Thank you for your answer. Is the code for presenting the experimental results with TensorboardX in basicSR or did you write it yourself?

Carefully read the readme.md would help I think~

**Comment by 924175302** [comment]
@IceClear
Hello, I read the readme.md carefully，I found out there used to be a file called logger.py .But now this file does not exist. It should be merged into util.py. So I still don't know how to run tensorboard. I didn't see a picture of PSNR when I was training

**Comment by IceClear** [comment]
> @IceClear
> Hello, I read the readme.md carefully，I found out there used to be a file called logger.py .But now this file does not exist. It should be merged into util.py. So I still don't know how to run tensorboard. I didn't see a picture of PSNR when I was training
Tensorboard is just a tool provided by tensorflow. All you need to do is just google it or read the document from the tensorflow official website.

**Comment by Slayerxxx** [comment]
> @xinntao Thanks for your reply! Actually, I train on my own dataset. I first pretrain G with l1 loss as you suggested. I found the loss decrease very fast, usually after 5k, the loss do not decrease obviously anymore. Then I turn to GAN training. I keep the weight the same as ESRGAN, since I do not know how to set it. I tried to pretrain D for 1k iterations, but it does not work. Now, my loss is as the figure shows below. I only train on the Y channel and as you can see, the val image is full of noise. I t seems the G generate noise and add it to the origin frame. Now, I mainly have two concerns:
> (1) Will some GAN improvement work? Such as label smoothing or any other tricks; Or if I switch to WGAN-GP instead of RAGAN ?
> (2) From the loss, I found at early iterations, the G loss has a slight decrease but become to increase after some iters and the D loss seems to shock and can be 0 sometimes. Does it mean the G is hard to fool the D and the training fails? Maybe the problem is my dataset? I am training on a video enhancement dataset in paper Multi-Frame Quality Enhancement for Compressed Video. So the GT is too hard for the G to learn?
> Thanks a lot for your patient reply.
> ![image](https://user-images.githubusercontent.com/22350795/65007654-69a07f00-d939-11e9-81a1-d8dcc4608e24.png)
> ![image](https://user-images.githubusercontent.com/22350795/65007669-73c27d80-d939-11e9-8125-9328f0addbbd.png)
> ![image](https://user-images.githubusercontent.com/22350795/65008099-c7819680-d93a-11e9-8bb3-8268875ea4d6.png)
> (This is the test result after 60k iterations, I found the continue training only result in different noise format)

Hi @IceClear ! I wonder how did you train on Y channel using VGG based perceptual loss? VGG is pretrained on RGB images, so can I ask how did you get over this? Hope for your reply. Thanks!
